invokeai/app/invocations/controlnet_image_processors.py

```diff
@@ -483,30 +485,67 @@ class TileResamplerProcessorInvocation(ImageProcessorInvocation):
 
     # res: int = InputField(default=512, ge=0, le=1024, description="The pixel resolution for each tile")
     down_sampling_rate: float = InputField(default=1.0, ge=1.0, le=8.0, description="Down sampling rate")
+    mode: Literal["regular", "blur", "var", "super"] = InputField(
+        default="regular", description="The controlnet tile model being used"
+    )
+
+    def apply_gaussian_blur(self, image_np: np.ndarray[Any, Any], ksize: int = 5, sigmaX: float = 1.0):
+        if ksize % 2 == 0:
+            ksize += 1  # ksize must be odd
+        blurred_image = cv2.GaussianBlur(image_np, (ksize, ksize), sigmaX=sigmaX)
+        return blurred_image
+
+    def apply_guided_filter(self, image_np: np.ndarray[Any, Any], radius: int, eps: float, scale: int):
+        filter = FastGuidedFilter(image_np, radius, eps, scale)
+        return filter.filter(image_np)
+
+    # based off https://huggingface.co/TTPlanet/TTPLanet_SDXL_Controlnet_Tile_Realistic
+    def tile_resample(self, np_img: np.ndarray[Any, Any]):
+        height, width, _ = np_img.shape
+
+        if self.mode == "regular":
+            np_img = HWC3(np_img)
+            if self.down_sampling_rate < 1.1:
+                return np_img
 
-    # tile_resample copied from sd-webui-controlnet/scripts/processor.py
-    def tile_resample(
-        self,
-        np_img: np.ndarray,
-        res=512,  # never used?
-        down_sampling_rate=1.0,
-    ):
-        np_img = HWC3(np_img)
-        if down_sampling_rate < 1.1:
+            new_height = int(float(height) / float(self.down_sampling_rate))
+            new_width = int(float(width) / float(self.down_sampling_rate))
+            np_img = cv2.resize(np_img, (new_width, new_height), interpolation=cv2.INTER_AREA)
             return np_img
-        H, W, C = np_img.shape
-        H = int(float(H) / float(down_sampling_rate))
-        W = int(float(W) / float(down_sampling_rate))
-        np_img = cv2.resize(np_img, (W, H), interpolation=cv2.INTER_AREA)
+
+        ratio = np.sqrt(1024.0 * 1024.0 / (width * height))
+
+        resize_w, resize_h = int(width * ratio), int(height * ratio)
+
+        if self.mode == "super":
+            resize_w, resize_h = int(width * ratio) // 48 * 48, int(height * ratio) // 48 * 48
+
+        np_img = cv2.resize(np_img, (resize_w, resize_h))
+
+        if self.mode == "blur":
+            blur_strength = random.sample([i / 10.0 for i in range(10, 201, 2)], k=1)[0]
+            radius = random.sample([i for i in range(1, 40, 2)], k=1)[0]  # noqa: C416
+            eps = random.sample([i / 1000.0 for i in range(1, 101, 2)], k=1)[0]
+            scale_factor = random.sample([i / 10.0 for i in range(10, 181, 5)], k=1)[0]
+
+            if random.random() > 0.5:
+                np_img = self.apply_gaussian_blur(np_img, ksize=int(blur_strength), sigmaX=blur_strength / 2)
+
+            if random.random() > 0.5:
+                np_img = self.apply_guided_filter(np_img, radius, eps, int(scale_factor))
```

Comment:
Why non-deterministic? Should we make these two different modes?

Reply:
Also how this particular model seems to prefer it: https://huggingface.co/xinsir/controlnet-tile-sdxl-1.0